src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1486,26 +1486,27 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
     mon = mark.monitor();
     assert(mon != nullptr, "must have monitor");
     // this object has a heavyweight monitor
-    nWant = mon->contentions(); // # of threads contending for monitor
-    nWait = mon->waiters();     // # of threads in Object.wait()
-    wantList =  Threads::get_pending_threads(tlh.list(), nWant + nWait, (address)mon);
+    nWant = mon->contentions(); // # of threads contending for monitor entry, but not re-entry
+    nWait = mon->waiters();     // # of threads waiting for notification,
+                                // or to re-enter monitor, in Object.wait()
+
+    // Get the actual set of threads trying to enter, or re-enter, the monitor.
+    wantList = Threads::get_pending_threads(tlh.list(), nWant + nWait, (address)mon);
     nWant = wantList->length();
   } else {
     // this object has a lightweight monitor
   }
 
   if (mon != nullptr) {
-    ObjectWaiter *waiter = mon->first_waiter();
-    for (int i = 0; i < nWait; i++) {
-      if (waiter == nullptr || (i != 0 && waiter == mon->first_waiter())) {
-        // robustness: the waiting list has gotten smaller
-        // The nWait count we got from the mon->waiters() may include the re-entering
-        // the monitor threads after being notified. Here we are correcting the actual
-        // number of the waiting threads by excluding those re-entering the monitor.
-        nWait = i;
-        break;
-      }
-      waiter = mon->next_waiter(waiter);
+    // Robustness: the actual waiting list can be smaller.
+    // The nWait count we got from the mon->waiters() may include the re-entering
+    // the monitor threads after being notified. Here we are correcting the actual
+    // number of the waiting threads by excluding those re-entering the monitor.
+    nWait = 0;
+    for (ObjectWaiter* waiter = mon->first_waiter();
+         waiter != nullptr && (nWait == 0 || waiter != mon->first_waiter());
```

Comment:
Sorry I do not understand the logic of this line. `waiters` is just a linked-list of `ObjectWaiter`s so all we need to do is traverse it and count the number of elements.

Reply:
The loop is endless without this extra condition, so we are getting a test execution timeout.
The `waiters` seems to be `circular doubly linked list` as we can see below:
```
inline void ObjectMonitor::AddWaiter(ObjectWaiter* node) {
  assert(node != nullptr, "should not add null node");
  assert(node->_prev == nullptr, "node already in list");
  assert(node->_next == nullptr, "node already in list");
  // put node at end of queue (circular doubly linked list)
  if (_WaitSet == nullptr) {
    _WaitSet = node;
    node->_prev = node;
    node->_next = node;
  } else {
    ObjectWaiter* head = _WaitSet;
    ObjectWaiter* tail = head->_prev;
    assert(tail->_next == head, "invariant check");
    tail->_next = node;
    head->_prev = node;
    node->_next = head;
    node->_prev = tail;
  }
}
```
I'll make sure nothing is missed and check the old code again.

Reply:
Okay, thanks.